### PR TITLE
manual clustering of contributions with same coordinates

### LINKED ISSUE #18: Enable Clustering of Map Markers in the Tree Profile Widget
- [x] Implement Clustering of Contributions
- [x] Sum total trees to form a cluster

### PR DESCRIPTION
**Pull request by norbertschuler**
Fixes #18

Changes in this pull request:
- this just sums up contributions with the same coordinates

### CODE REVIEW

File: src/TreeMap/App.svelte [hunk 1]
```diff
@@ -80,12 +80,28 @@
 
         map.on("load", () => {
             fetchContributionsData.then((contributions) => {
+                let unique_coordinates = [];
+                let unique_coordinates_count = [];
+                contributions = Object.values(contributions).filter((c) => {
+                  if (c.geometry && c.geometry.type === 'Point' && c.geometry.coordinates) {
+                    if (unique_coordinates_count[c.geometry.coordinates.toString()]) {
+                      unique_coordinates_count[c.geometry.coordinates.toString()] += Number(c.properties.treeCount);
```
Comment (norbertschuler) [inline src/TreeMap/App.svelte:88]: Collect the sum of the treeCount values for unique coordinates in this associative array.

File: src/TreeMap/App.svelte [hunk 2]
```diff
@@ -80,12 +80,28 @@
...
+                  }
+                  return false;
+                });
                 if (contributions.length > 0) {
                     contributions.map((contribution) => {
                         if (contribution.geometry) {
                             var el = document.createElement("div");
                             var treeCount = document.createTextNode(
-                                contribution.properties.treeCount
+                                getFormattedNumber(unique_coordinates_count[contribution.geometry.coordinates.toString()],locale)
```
Comment (norbertschuler) [inline src/TreeMap/App.svelte:104]: Override the treeCount value of this contributions with the summed up value from the associative array for every unique coordinates.

File: src/TreeMap/App.svelte [hunk 3]
```diff
@@ -80,12 +80,28 @@
...
+                let unique_coordinates = [];
+                let unique_coordinates_count = [];
+                contributions = Object.values(contributions).filter((c) => {
+                  if (c.geometry && c.geometry.type === 'Point' && c.geometry.coordinates) {
+                    if (unique_coordinates_count[c.geometry.coordinates.toString()]) {
+                      unique_coordinates_count[c.geometry.coordinates.toString()] += Number(c.properties.treeCount);
+                    } else {
+                      unique_coordinates_count[c.geometry.coordinates.toString()] = Number(c.properties.treeCount);
+                    }
+                    if (!unique_coordinates.includes(c.geometry.coordinates.toString())) {
```
Comment (norbertschuler) [inline src/TreeMap/App.svelte:92]: Only keep one contribution item for unique coordinates.

**Comment by sagararyal** [comment]
@norbertschuler Interesting experiment, but we need to cluster by region, and be able to disable cluster at different zoom level.

With the approach above, cases like below: where someone has planted many trees are not clustered:
Screen Shot 2021-02-02 at 1 48 31 PM

So 
We should use the normal cluster behavior; but instead of aggregating the count of items in a cluster, we need to aggregate the properties of trees planted.

**Comment by norbertschuler** [comment]
> We should use the normal cluster behavior; but instead of aggregating the count of items in a cluster, we need to aggregate the properties of trees planted.

O.k., then someone needs to implement Mapbox clustering (maybe also for the My Forest map in the Planet-Webapp) as I mentioned here: https://github.com/Plant-for-the-Planet-org/planet-widgets/issues/18#issuecomment-771538301

**Pull request closed without merging**